File: linked_list/linked_list/Y_intersect.c
#include <stdio.h>
#include <stdlib.h>

struct Node {
    int val;
    struct Node* next;
};

struct Node* newNode(int v) {
    struct Node* n = (struct Node*)malloc(sizeof(struct Node));
    n->val = v;
    n->next = NULL;
    return n;
}

void show(struct Node* h, char* name) {
    printf("%s: ", name);
    struct Node* cur = h;
    while (cur) {
        printf("%d -> ", cur->val);
        cur = cur->next;
    }
    printf("NULL\n");
}

int length(struct Node* h) {
    int l = 0;
    while (h) {
        l++;
        h = h->next;
    }
    return l;
}

struct Node* intersect(struct Node* f, struct Node* s) {
    int lf = length(f), ls = length(s);
    if (lf > ls) {
        for (int i = 0; i < lf - ls; i++) f = f->next;
    } else {
        for (int i = 0; i < ls - lf; i++) s = s->next;
    }
    while (f && s) {
        if (f == s) return f;
        f = f->next;
        s = s->next;
    }
    return NULL;
}

int main() {
    struct Node* c = newNode(30);
    c->next = newNode(40);
    c->next->next = newNode(50);

    struct Node* f = newNode(10);
    f->next = newNode(20);
    f->next->next = c;

    struct Node* s = newNode(15);
    s->next = c;

    show(f, "First");
    show(s, "Second");

    struct Node* x = intersect(f, s);
    if (x) printf("Intersect at: %d\n", x->val);
    else   printf("No intersection\n");

    return 0;
}

File: linked_list/linked_list/Y_intersect.py
class Node:
    def __init__(self, val):
        self.val = val
        self.next = None


def show(head, name):
    cur = head
    print(name, end=": ")
    while cur:
        print(cur.val, end=" -> ")
        cur = cur.next
    print("None")


def length(head):
    l = 0
    cur = head
    while cur:
        l += 1
        cur = cur.next
    return l


def intersect(f, s):
    lf, ls = length(f), length(s)

    if lf > ls:
        for _ in range(lf - ls):
            f = f.next
    else:
        for _ in range(ls - lf):
            s = s.next

    while f and s:
        if f is s:
            return f
        f, s = f.next, s.next
    return None


c = Node(30)
c.next = Node(40)
c.next.next = Node(50)

f = Node(10)
f.next = Node(20)
f.next.next = c

s = Node(15)
s.next = c

show(f, "First")
show(s, "Second")

x = intersect(f, s)
if x:
    print("Intersect at:", x.val)
else:
    print("No intersection")
